src/app/clusters/bindings/bindings.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/****************************************************************************
+ * @file
+ * @brief Implementation for the Binding Server Cluster
+ ***************************************************************************/
+
+#include "af.h"
+
+#include <app/util/binding-table.h>
+#include <support/logging/CHIPLogging.h>
+
+using namespace chip;
+
+EmberStatus prepareBinding(EmberBindingTableEntry & binding, NodeId nodeId, GroupId groupId, EndpointId endpointId,
+                           ClusterId clusterId)
+{
+    if (groupId && nodeId)
+    {
+        return EMBER_BAD_ARGUMENT;
+    }
+
+    if (groupId)
```

Comment:
the if/else blocks seem to contain similar code. Could we only take out differences?

like:

```cpp
binding.clusterId = clusterId;
binding.local     = emberAfCurrentCommand()->apsFrame->destinationEndpoint;

if (groupId != 0) { // should we have a constant of kEmptyGroupId? 
  binding.type      = EMBER_MULTICAST_BINDING;
  binding.groupId   = groupId;
} else {
  binding.type      = EMBER_UNICAST_BINDING;
  binding.nodeId    = nodeId;
  binding.remote    = endpointId;
}
```

followup question:
  - should we clear bindig parameters unused (e.g. nodeid ofr multicast, groupid for unicast)?

Reply:
> the if/else blocks seem to contain similar code. Could we only take out differences?
> 
> like:
> 
> ```c++
> binding.clusterId = clusterId;
> binding.local     = emberAfCurrentCommand()->apsFrame->destinationEndpoint;
> 
> if (groupId != 0) { // should we have a constant of kEmptyGroupId? 
>   binding.type      = EMBER_MULTICAST_BINDING;
>   binding.groupId   = groupId;
> } else {
>   binding.type      = EMBER_UNICAST_BINDING;
>   binding.nodeId    = nodeId;
>   binding.remote    = endpointId;
> }

Done.

> ```
> 
> followup question:
> 
> * should we clear bindig parameters unused (e.g. nodeid ofr multicast, groupid for unicast)?

`nodeId` and `groupId` are members of an union: https://github.com/project-chip/connectedhomeip/blob/6602287f7faa3030aefcd4ba00c61b7eab74c382/src/app/util/types_stub.h#L546